File: md/iron-condor/tools/scrape_page.mjs
#!/usr/bin/env node
/**
 * Scrape a rendered page via CDP, convert to markdown via defuddle.
 *
 * Dependencies:
 *   npm install ws jsdom defuddle
 *
 * Usage:
 *   node tools/scrape_page.mjs <url> [output.md]
 *   node tools/scrape_page.mjs <url>              # print to stdout
 *   node tools/scrape_page.mjs <url> -w ws://... # custom CDP WebSocket URL
 *   node tools/scrape_page.mjs <url> -t 5000     # wait 5s for JS rendering (default 3000)
 */

import { WebSocket } from "ws";
import { JSDOM } from "jsdom";
import { Defuddle } from "defuddle/node";
import { setTimeout as sleep } from "timers/promises";
import { writeFileSync, unlinkSync } from "fs";
import { tmpdir } from "os";
import { join } from "path";

const DEFAULT_RENDER_TIMEOUT = 3000;

/**
 * Find the first Chrome CDP websocket URL, skipping DevTools itself.
 */
function findChromeDebuggerUrl(targets) {
  for (const target of targets) {
    if (target.url.startsWith("devtools://")) continue;
    if (target.webSocketDebuggerUrl) return target.webSocketDebuggerUrl;
  }
  return null;
}

/**
 * Send a CDP command and wait for its response by matching ID.
 */
async function sendAndRecv(ws, id, method, params = {}) {
  const cmd = { id, method, params };
  ws.send(JSON.stringify(cmd));

  return new Promise((resolve) => {
    ws.on("message", (data) => {
      const msg = JSON.parse(data.toString());
      if (msg.id === id) resolve(msg);
    });
  });
}

/**
 * Navigate to a URL via CDP and return document.body.outerHTML.
 */
async function navigateAndGetHtml(ws, url, renderTimeout) {
  // Enable Page domain
  await sendAndRecv(ws, 1, "Page.enable");

  // Navigate
  await sendAndRecv(ws, 2, "Page.navigate", { url });

  // Wait for load event
  await new Promise((resolve) => {
    ws.on("message", (data) => {
      const msg = JSON.parse(data.toString());
      if (msg.method === "Page.loadEventFired") resolve();
    });
  });

  // Extra wait for JS rendering
  await sleep(renderTimeout);

  // Get document
  const docResp = await sendAndRecv(ws, 3, "DOM.getDocument", { depth: 0 });
  const rootNodeId = docResp.result.root.nodeId;

  // QuerySelector for body
  const bodyResp = await sendAndRecv(ws, 4, "DOM.querySelector", {
    nodeId: rootNodeId,
    selector: "body",
  });
  const bodyNodeId = bodyResp.result.nodeId;

  // Get outerHTML
  const htmlResp = await sendAndRecv(ws, 5, "DOM.getOuterHTML", {
    nodeId: bodyNodeId,
  });
  return htmlResp.result.outerHTML;
}

/**
 * Convert HTML to markdown using defuddle.
 */
async function htmlToMarkdown(html, baseUrl) {
  const dom = new JSDOM(html, { url: baseUrl });
  const result = await Defuddle(dom.window.document, baseUrl, {
    markdown: true,
  });
  return result.content || "";
}

// ---------------------------------------------------------------------------

function parseArgs(argv) {
  const args = {
    url: null,
    outputPath: null,
    wsUrl: null,
    renderTimeout: DEFAULT_RENDER_TIMEOUT,
  };

  for (let i = 2; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === "-w" || arg === "--ws-url") {
      args.wsUrl = argv[++i];
    } else if (arg === "-t" || arg === "--timeout") {
      args.renderTimeout = parseInt(argv[++i], 10);
    } else if (arg === "-h" || arg === "--help") {
      args.help = true;
    } else if (!args.url) {
      args.url = arg;
    } else if (!args.outputPath) {
      args.outputPath = arg;
    }
  }

  return args;
}

function printUsage() {
  console.log(`Usage: node scrape_page.mjs <url> [output.md] [options]
Options:
  -w, --ws-url <url>   CDP WebSocket URL (auto-detected by default)
  -t, --timeout <ms>  Wait time for JS rendering (default: ${DEFAULT_RENDER_TIMEOUT})
  -h, --help           Show this help message
Examples:
  node scrape_page.mjs "https://example.com"
  node scrape_page.mjs "https://example.com" "output.md"
  node tools/scrape_page.mjs "https://example.com" -t 5000
`);
}

async function main() {
  const args = parseArgs(process.argv);

  if (args.help || !args.url) {
    printUsage();
    process.exit(args.help ? 0 : 1);
    return;
  }

  // Determine CDP WebSocket URL
  let targets;
  let autoWsUrl;
  try {
    const resp = await fetch("http://localhost:9222/json");
    targets = await resp.json();
    autoWsUrl = findChromeDebuggerUrl(targets);
  } catch {
    console.error(
      "Error: Could not connect to Chrome debugger at http://localhost:9222/json"
    );
    console.error(
      "Is Chrome running with --remote-debugging-port=9222?"
    );
    process.exit(1);
    return;
  }

  const finalWsUrl = args.wsUrl || autoWsUrl;

  if (!finalWsUrl) {
    console.error(
      "Error: Could not find a Chrome debugger page."
    );
    process.exit(1);
    return;
  }

  console.error(`Connecting to Chrome at ${finalWsUrl}`);

  // Connect via WebSocket
  const ws = new WebSocket(finalWsUrl, {
    maxPayload: 200 * 1024 * 1024,
  });

  await new Promise((resolve, reject) => {
    ws.on("open", resolve);
    ws.on("error", reject);
  });

  const html = await navigateAndGetHtml(ws, args.url, args.renderTimeout);
  ws.close();

  // Save HTML to temp file (for potential debug) — only if we need it
  const htmlPath = join(tmpdir(), `scrape_${Date.now()}.html`);
  writeFileSync(htmlPath, html, "utf8");
  console.error(`HTML saved to: ${htmlPath}`);

  // Convert to markdown
  const markdown = await htmlToMarkdown(html, args.url);

  if (args.outputPath) {
    writeFileSync(args.outputPath, markdown, "utf8");
    console.error(`Markdown saved to: ${args.outputPath}`);
  } else {
    console.log(markdown);
  }

  // Cleanup temp HTML
  unlinkSync(htmlPath);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: md/iron-condor/tools/scrape_page.py
#!/usr/bin/env python3
"""
Scrape a rendered page via CDP, save HTML to temp, convert to markdown via defuddle.

Usage:
    python tools/scrape_page.py "https://example.com" "output.md"
"""

import argparse
import asyncio
import json
import os
import subprocess
import sys
import tempfile
import urllib.request
import websockets


def find_chrome_debugger_url() -> str | None:
    """Get the first available Chrome CDP websocket URL (skip DevTools itself)."""
    try:
        with urllib.request.urlopen("http://localhost:9222/json", timeout=5) as resp:
            targets = json.loads(resp.read())
        for target in targets:
            url = target.get("url", "")
            if url.startswith("devtools://"):
                continue
            if ws_url := target.get("webSocketDebuggerUrl"):
                return ws_url
    except Exception:
        pass
    return None


async def send_and_recv(ws, cmd: dict) -> dict:
    """Send a CDP command and wait for its response by matching ID."""
    await ws.send(json.dumps(cmd))
    while True:
        msg = json.loads(await ws.recv())
        if msg.get("id") == cmd["id"]:
            return msg


async def navigate_and_get_html(ws_url: str, url: str) -> str:
    """Navigate to URL and get body outerHTML after page load."""
    async with websockets.connect(ws_url, max_size=None) as ws:
        # Enable Page domain
        await send_and_recv(ws, {"id": 1, "method": "Page.enable"})

        # Navigate
        await send_and_recv(ws, {
            "id": 2,
            "method": "Page.navigate",
            "params": {"url": url}
        })

        # Wait for load event
        while True:
            msg = json.loads(await ws.recv())
            if msg.get("method") == "Page.loadEventFired":
                break

        # Wait extra for JS rendering
        await asyncio.sleep(3)

        # Get document
        resp = await send_and_recv(ws, {
            "id": 3,
            "method": "DOM.getDocument",
            "params": {"depth": 0}
        })
        root_node_id = resp["result"]["root"]["nodeId"]

        # QuerySelector for body
        resp = await send_and_recv(ws, {
            "id": 4,
            "method": "DOM.querySelector",
            "params": {"nodeId": root_node_id, "selector": "body"}
        })
        body_node_id = resp["result"]["nodeId"]

        # Get outerHTML
        resp = await send_and_recv(ws, {
            "id": 5,
            "method": "DOM.getOuterHTML",
            "params": {"nodeId": body_node_id}
        })
        return resp["result"]["outerHTML"]


def run_defuddle(html_path: str, output_path: str):
    """Convert HTML to markdown using defuddle."""
    cmd = [
        "npx", "defuddle", "parse",
        html_path,
        "--markdown",
        "-o", output_path
    ]
    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        print(f"defuddle error: {result.stderr}", file=sys.stderr)
        sys.exit(1)
    print(f"Saved markdown to: {output_path}")


def main():
    parser = argparse.ArgumentParser(description="Scrape rendered page via CDP, save as markdown")
    parser.add_argument("url", help="URL to scrape")
    parser.add_argument("output", help="Output markdown path")
    args = parser.parse_args()

    ws_url = find_chrome_debugger_url()
    if not ws_url:
        print("Error: Could not find Chrome debugger. Is Chrome running with --remote-debugging-port=9222?", file=sys.stderr)
        sys.exit(1)

    print(f"Connecting to Chrome at {ws_url}")
    html = asyncio.run(navigate_and_get_html(ws_url, args.url))

    # Save HTML to temp file
    with tempfile.NamedTemporaryFile(mode="w", suffix=".html", delete=False) as f:
        f.write(html)
        html_path = f.name

    print(f"Saved HTML to: {html_path}")

    # Convert to markdown
    run_defuddle(html_path, args.output)

    # Cleanup temp HTML
    os.unlink(html_path)


if __name__ == "__main__":
    main()
